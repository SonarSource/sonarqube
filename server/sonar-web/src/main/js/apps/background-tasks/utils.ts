/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { ONE_SECOND } from '../../helpers/constants';
import { toShortISO8601String } from '../../helpers/dates';
import { ActivityRequestParameters, Task, TaskStatuses } from '../../types/tasks';
import { ALL_TYPES, CURRENTS, STATUSES } from './constants';

export interface Query {
  currents: string;
  maxExecutedAt?: Date;
  minSubmittedAt?: Date;
  query: string;
  status: string;
  taskType: string;
}

export function updateTask(tasks: Task[], newTask: Task) {
  return tasks.map((task) => (task.id === newTask.id ? newTask : task));
}

export function mapFiltersToParameters(filters: Partial<Query> = {}) {
  const parameters: ActivityRequestParameters = {};

  if (filters.status === STATUSES.ALL) {
    parameters.status = [
      TaskStatuses.Pending,
      TaskStatuses.InProgress,
      TaskStatuses.Success,
      TaskStatuses.Failed,
      TaskStatuses.Canceled,
    ].join();
  } else if (filters.status === STATUSES.ALL_EXCEPT_PENDING) {
    parameters.status = [
      TaskStatuses.InProgress,
      TaskStatuses.Success,
      TaskStatuses.Failed,
      TaskStatuses.Canceled,
    ].join();
  } else {
    parameters.status = filters.status;
  }

  if (filters.taskType !== ALL_TYPES) {
    parameters.type = filters.taskType;
  }

  if (filters.currents !== CURRENTS.ALL) {
    parameters.onlyCurrents = true;
  }

  if (filters.minSubmittedAt) {
    parameters.minSubmittedAt = toShortISO8601String(filters.minSubmittedAt);
  }

  if (filters.maxExecutedAt) {
    parameters.maxExecutedAt = toShortISO8601String(filters.maxExecutedAt);
  }

  if (filters.query) {
    parameters.q = filters.query;
  }

  return parameters;
}

const ONE_MINUTE = 60 * ONE_SECOND;
const ONE_HOUR = 60 * ONE_MINUTE;

function format(int: number, suffix: string) {
  return `${int}${suffix}`;
}

export function formatDuration(value: number | undefined) {
  if (!value) {
    return '';
  }
  if (value < ONE_SECOND) {
    return format(value, 'ms');
  } else if (value < ONE_SECOND * 10) {
    const seconds = Math.floor(value / ONE_SECOND);
    const ms = value - seconds * ONE_SECOND;
    return seconds + '.' + format(ms, 's');
  } else if (value < ONE_MINUTE) {
    const seconds = Math.floor(value / ONE_SECOND);
    return format(seconds, 's');
  } else if (value < ONE_MINUTE * 10) {
    const minutes = Math.floor(value / ONE_MINUTE);
    const seconds = Math.floor((value - minutes * ONE_MINUTE) / ONE_SECOND);
    return format(minutes, 'min') + ' ' + format(seconds, 's');
  }
  const hours = Math.floor(value / ONE_HOUR);
  const minutes = Math.floor((value - hours * ONE_HOUR) / ONE_MINUTE);
  return format(hours, 'h') + ' ' + format(minutes, 'min');
}
