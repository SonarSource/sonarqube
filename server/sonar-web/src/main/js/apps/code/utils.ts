/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { isPullRequest } from '~sonar-aligned/helpers/branch-like';
import { isPortfolioLike } from '~sonar-aligned/helpers/component';
import { ComponentQualifier } from '~sonar-aligned/types/component';
import { MetricKey } from '~sonar-aligned/types/metrics';
import { CCT_SOFTWARE_QUALITY_METRICS, OLD_TAXONOMY_METRICS } from '../../helpers/constants';
import { BranchLike } from '../../types/branch-like';

const METRICS = [
  MetricKey.ncloc,
  ...CCT_SOFTWARE_QUALITY_METRICS,
  ...OLD_TAXONOMY_METRICS,
  MetricKey.security_hotspots,
  MetricKey.coverage,
  MetricKey.duplicated_lines_density,
];

const APPLICATION_METRICS = [MetricKey.alert_status, ...METRICS];

const PORTFOLIO_METRICS = [
  MetricKey.releasability_rating,
  MetricKey.security_rating,
  MetricKey.software_quality_security_rating,
  MetricKey.reliability_rating,
  MetricKey.software_quality_reliability_rating,
  MetricKey.sqale_rating,
  MetricKey.software_quality_maintainability_rating,
  MetricKey.security_review_rating,
  MetricKey.ncloc,
];

const NEW_PORTFOLIO_METRICS = [
  MetricKey.releasability_rating,
  MetricKey.new_security_rating,
  MetricKey.new_software_quality_security_rating,
  MetricKey.new_reliability_rating,
  MetricKey.new_software_quality_reliability_rating,
  MetricKey.new_maintainability_rating,
  MetricKey.new_software_quality_maintainability_rating,
  MetricKey.new_security_review_rating,
  MetricKey.new_lines,
];

const LEAK_METRICS = [
  MetricKey.new_lines,
  ...CCT_SOFTWARE_QUALITY_METRICS,
  ...OLD_TAXONOMY_METRICS,
  MetricKey.security_hotspots,
  MetricKey.new_coverage,
  MetricKey.new_duplicated_lines_density,
];

export function getCodeMetrics(
  qualifier: string,
  branchLike?: BranchLike,
  options: { includeQGStatus?: boolean; newCode?: boolean } = {},
) {
  if (isPortfolioLike(qualifier)) {
    let metrics: MetricKey[] = [];
    if (options?.newCode === undefined) {
      metrics = [...NEW_PORTFOLIO_METRICS, ...PORTFOLIO_METRICS];
    } else if (options?.newCode) {
      metrics = [...NEW_PORTFOLIO_METRICS];
    } else {
      metrics = [...PORTFOLIO_METRICS];
    }
    return options.includeQGStatus ? metrics.concat(MetricKey.alert_status) : metrics;
  }
  if (qualifier === ComponentQualifier.Application) {
    return [...APPLICATION_METRICS];
  }
  if (isPullRequest(branchLike)) {
    return [...LEAK_METRICS];
  }
  return [...METRICS];
}

export function mostCommonPrefix(strings: string[]) {
  const sortedStrings = strings.slice(0).sort((a, b) => a.localeCompare(b));
  const firstString = sortedStrings[0];
  const firstStringLength = firstString.length;
  const lastString = sortedStrings[sortedStrings.length - 1];
  let i = 0;
  while (i < firstStringLength && firstString.charAt(i) === lastString.charAt(i)) {
    i++;
  }
  const prefix = firstString.slice(0, i);
  const prefixTokens = prefix.split(/[\s\\/]/);
  const lastPrefixPart = prefixTokens[prefixTokens.length - 1];
  return prefix.slice(0, prefix.length - lastPrefixPart.length);
}
