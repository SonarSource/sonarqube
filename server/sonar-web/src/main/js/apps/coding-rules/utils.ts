/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { omit } from 'lodash';
import { RestRuleDetails, RuleDetails } from '../../types/types';

export const REST_RULE_KEYS_TO_OLD_KEYS = {
  repositoryKey: 'repo',
  external: 'isExternal',
  markdownDescription: 'mdDesc',
  markdownNote: 'mdNote',
  template: 'isTemplate',
  systemTags: 'sysTags',
  language: 'lang',
  languageName: 'langName',
};

// Mapping new resource to old api. We should get rid of it with migration of rules/search & rule/details
export function mapRestRuleToRule(rule: RestRuleDetails): RuleDetails {
  return {
    ...omit(rule, Object.keys(REST_RULE_KEYS_TO_OLD_KEYS)),
    ...Object.entries(REST_RULE_KEYS_TO_OLD_KEYS).reduce(
      (obj, [key, value]: [keyof RestRuleDetails, keyof RuleDetails]) => {
        obj[value] = rule[key] as never;
        return obj;
      },
      {} as RuleDetails,
    ),
    params: rule.parameters?.map((param) => ({
      ...omit(param, 'htmlDescription'),
      htmlDesc: param.htmlDescription,
    })),
  };
}
