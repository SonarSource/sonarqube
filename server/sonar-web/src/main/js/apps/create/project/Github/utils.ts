/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { isWebUri } from 'valid-url';
import { getGithubClientId } from '../../../../api/alm-integrations';
import { getHostUrl } from '../../../../helpers/urls';
import { AlmKeys } from '../../../../types/alm-settings';
import { DopSetting } from '../../../../types/dop-translation';

export async function redirectToGithub(params: {
  isMonorepoSetup: boolean;
  selectedDopSetting?: DopSetting;
}) {
  const { isMonorepoSetup, selectedDopSetting } = params;

  if (selectedDopSetting?.url === undefined) {
    return;
  }

  const { clientId } = await getGithubClientId(selectedDopSetting.key);

  if (clientId === undefined) {
    throw new Error('Received no GitHub client id');
  }
  const queryParams = [
    { param: 'client_id', value: clientId },
    {
      param: 'redirect_uri',
      value: encodeURIComponent(
        `${getHostUrl()}/projects/create?mode=${AlmKeys.GitHub}&dopSetting=${
          selectedDopSetting.id
        }${isMonorepoSetup ? '&mono=true' : ''}`,
      ),
    },
  ]
    .map(({ param, value }) => `${param}=${value}`)
    .join('&');

  let instanceRootUrl;
  // Strip the api section from the url, since we're not hitting the api here.
  if (selectedDopSetting.url.includes('/api/v3')) {
    // GitHub Enterprise
    instanceRootUrl = selectedDopSetting.url.replace('/api/v3', '');
  } else {
    // github.com
    instanceRootUrl = selectedDopSetting.url.replace('api.', '');
  }

  // strip the trailing /
  instanceRootUrl = instanceRootUrl.replace(/\/$/, '');
  if (isWebUri(instanceRootUrl) === undefined) {
    throw new Error('Invalid GitHub URL');
  } else {
    window.location.replace(`${instanceRootUrl}/login/oauth/authorize?${queryParams}`);
  }
}
