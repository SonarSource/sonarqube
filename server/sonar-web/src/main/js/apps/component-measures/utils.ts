/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { groupBy, memoize, sortBy, toPairs } from 'lodash';
import { isBranch, isPullRequest } from '~sonar-aligned/helpers/branch-like';
import { ComponentQualifier } from '~sonar-aligned/types/component';
import { MetricKey, MetricType } from '~sonar-aligned/types/metrics';
import { RawQuery } from '~sonar-aligned/types/router';
import { enhanceMeasure } from '../../components/measure/utils';
import {
  CCT_SOFTWARE_QUALITY_METRICS,
  HIDDEN_METRICS,
  LEAK_CCT_SOFTWARE_QUALITY_METRICS,
  LEAK_OLD_TAXONOMY_METRICS,
  LEAK_OLD_TAXONOMY_RATINGS,
  OLD_TAXONOMY_METRICS,
  OLD_TAXONOMY_RATINGS,
  SOFTWARE_QUALITY_RATING_METRICS,
} from '../../helpers/constants';
import { getLocalizedMetricName, translate } from '../../helpers/l10n';
import {
  areCCTMeasuresComputed,
  areLeakCCTMeasuresComputed,
  areSoftwareQualityRatingsComputed,
  getDisplayMetrics,
  isDiffMetric,
  MEASURES_REDIRECTION,
} from '../../helpers/measures';
import {
  cleanQuery,
  parseAsOptionalBoolean,
  parseAsString,
  serializeString,
} from '../../helpers/query';
import { BranchLike } from '../../types/branch-like';
import { Domain, MeasurePageView } from '../../types/measures';
import {
  ComponentMeasure,
  ComponentMeasureEnhanced,
  Dict,
  Measure,
  MeasureEnhanced,
  Metric,
} from '../../types/types';
import { BubblesByDomain } from './config/bubbles';
import { domains } from './config/domains';

export const BUBBLES_FETCH_LIMIT = 50;
export const PROJECT_OVERVIEW = 'project_overview';
export const DEFAULT_VIEW = MeasurePageView.tree;
export const DEFAULT_METRIC = PROJECT_OVERVIEW;
export const KNOWN_DOMAINS = [
  'Releasability',
  'Security',
  'Reliability',
  'Maintainability',
  'SecurityReview',
  'Coverage',
  'Duplications',
  'Size',
  'Complexity',
];

const DEPRECATED_DOMAINS_METRICS = [MetricKey.high_impact_accepted_issues];

const HIDDEN_DOMAIN_METRICS = [
  ...DEPRECATED_DOMAINS_METRICS,
  MetricKey.blocker_violations,
  MetricKey.new_blocker_violations,
  MetricKey.critical_violations,
  MetricKey.new_critical_violations,
  MetricKey.major_violations,
  MetricKey.new_major_violations,
  MetricKey.info_violations,
  MetricKey.new_info_violations,
  MetricKey.minor_violations,
  MetricKey.new_minor_violations,
  MetricKey.software_quality_blocker_issues,
  MetricKey.new_software_quality_blocker_issues,
  MetricKey.software_quality_high_issues,
  MetricKey.new_software_quality_high_issues,
  MetricKey.software_quality_medium_issues,
  MetricKey.new_software_quality_medium_issues,
  MetricKey.software_quality_low_issues,
  MetricKey.new_software_quality_low_issues,
  MetricKey.software_quality_info_issues,
  MetricKey.new_software_quality_info_issues,
];

const ISSUES_METRICS = [
  MetricKey.accepted_issues,
  MetricKey.new_accepted_issues,
  MetricKey.confirmed_issues,
  MetricKey.false_positive_issues,
  MetricKey.violations,
  MetricKey.new_violations,
];

export const populateDomainsFromMeasures = memoize(
  (measures: MeasureEnhanced[], isStandardMode = false): Domain[] => {
    let populatedMeasures = measures
      .filter((measure) => !HIDDEN_DOMAIN_METRICS.includes(measure.metric.key as MetricKey))
      .map((measure) => {
        const isDiff = isDiffMetric(measure.metric.key);
        const calculatedValue = isDiff ? measure.leak : measure.value;

        return {
          ...measure,
          ...{ [isDiff ? 'leak' : 'value']: calculatedValue },
        };
      });

    if (!isStandardMode && areLeakCCTMeasuresComputed(measures)) {
      populatedMeasures = populatedMeasures.filter(
        (measure) => !LEAK_OLD_TAXONOMY_METRICS.includes(measure.metric.key as MetricKey),
      );
    } else {
      populatedMeasures = populatedMeasures.filter(
        (measure) => !LEAK_CCT_SOFTWARE_QUALITY_METRICS.includes(measure.metric.key as MetricKey),
      );
    }

    // Both new and overall code will exist after next analysis
    if (!isStandardMode && areSoftwareQualityRatingsComputed(measures)) {
      populatedMeasures = populatedMeasures.filter(
        (measure) =>
          !OLD_TAXONOMY_RATINGS.includes(measure.metric.key as MetricKey) &&
          !LEAK_OLD_TAXONOMY_RATINGS.includes(measure.metric.key as MetricKey),
      );
    } else {
      populatedMeasures = populatedMeasures.filter(
        (measure) => !SOFTWARE_QUALITY_RATING_METRICS.includes(measure.metric.key as MetricKey),
      );
    }

    if (!isStandardMode && areCCTMeasuresComputed(measures)) {
      populatedMeasures = populatedMeasures.filter(
        (measure) => !OLD_TAXONOMY_METRICS.includes(measure.metric.key as MetricKey),
      );
    } else {
      populatedMeasures = populatedMeasures.filter(
        (measure) => !CCT_SOFTWARE_QUALITY_METRICS.includes(measure.metric.key as MetricKey),
      );
    }

    return groupByDomains(populatedMeasures);
  },
);

export function getMetricSubnavigationName(
  metric: Metric,
  translateFn: (metric: Metric) => string,
  isDiff = false,
  isStandardMode = false,
) {
  // MQR mode and old taxonomy metrics, we return "Issues" for them anyway
  if (!isStandardMode && OLD_TAXONOMY_METRICS.includes(metric.key as MetricKey)) {
    return translate('component_measures.awaiting_analysis.name');
  }
  if (!isStandardMode && LEAK_OLD_TAXONOMY_METRICS.includes(metric.key as MetricKey)) {
    return translate('component_measures.leak_awaiting_analysis.name');
  }

  if (
    [
      ...LEAK_CCT_SOFTWARE_QUALITY_METRICS,
      ...CCT_SOFTWARE_QUALITY_METRICS,
      ...ISSUES_METRICS,
      ...OLD_TAXONOMY_METRICS,
      ...LEAK_OLD_TAXONOMY_METRICS,
    ].includes(metric.key as MetricKey)
  ) {
    return translate(
      `component_measures.metric.${metric.key}.${isDiff ? 'detailed_name' : 'name'}`,
    );
  }
  return translateFn(metric);
}

export function filterMeasures(measures: MeasureEnhanced[]): MeasureEnhanced[] {
  return measures.filter((measure) => !HIDDEN_METRICS.includes(measure.metric.key as MetricKey));
}

export function sortMeasures(
  domainName: string,
  measures: Array<MeasureEnhanced | string>,
): Array<MeasureEnhanced | string> {
  const config = domains[domainName] || {};
  const configOrder = config.order || [];
  return sortBy(measures, [
    (item: MeasureEnhanced | string) => {
      if (typeof item === 'string') {
        return configOrder.indexOf(item);
      }
      const idx = configOrder.indexOf(item.metric.key);
      return idx >= 0 ? idx : configOrder.length;
    },
    (item: MeasureEnhanced | string) =>
      typeof item === 'string' ? item : getLocalizedMetricName(item.metric),
  ]);
}

export function addMeasureCategories(domainName: string, measures: MeasureEnhanced[]) {
  const categories = domains[domainName] && domains[domainName].categories;
  if (categories && categories.length > 0) {
    return [...categories, ...measures];
  }
  return measures;
}

export function enhanceComponent(
  component: ComponentMeasure,
  metric: Pick<Metric, 'key'> | undefined,
  metrics: Dict<Metric>,
): ComponentMeasureEnhanced {
  if (!component.measures) {
    return { ...component, measures: [] };
  }

  const enhancedMeasures = component.measures.map((measure) => enhanceMeasure(measure, metrics));
  const measure = metric && enhancedMeasures.find((measure) => measure.metric.key === metric.key);
  const value = measure && measure.value;
  const leak = measure && measure.leak;
  return { ...component, value, leak, measures: enhancedMeasures };
}

export function isSecurityReviewMetric(metricKey: MetricKey | string): boolean {
  return [
    MetricKey.security_hotspots,
    MetricKey.security_hotspots_reviewed,
    MetricKey.security_review_rating,
    MetricKey.new_security_hotspots,
    MetricKey.new_security_hotspots_reviewed,
    MetricKey.new_security_review_rating,
  ].includes(metricKey as MetricKey);
}

export function banQualityGateMeasure({ measures = [], qualifier }: ComponentMeasure): Measure[] {
  const bannedMetrics: string[] = [];
  if (ComponentQualifier.Portfolio !== qualifier && ComponentQualifier.SubPortfolio !== qualifier) {
    bannedMetrics.push(MetricKey.alert_status);
  }
  if (qualifier === ComponentQualifier.Application) {
    bannedMetrics.push(MetricKey.releasability_rating, MetricKey.releasability_effort);
  }
  return measures.filter((measure) => !bannedMetrics.includes(measure.metric));
}

export const groupByDomains = memoize((measures: MeasureEnhanced[]) => {
  const domains = toPairs(groupBy(measures, (measure) => measure.metric.domain)).map((r) => ({
    name: r[0],
    measures: r[1],
  }));

  return sortBy(domains, [
    (domain: { measures: MeasureEnhanced[]; name: string }) => {
      const idx = KNOWN_DOMAINS.indexOf(domain.name);
      return idx >= 0 ? idx : KNOWN_DOMAINS.length;
    },
    'name',
  ]);
});

export function hasList(metric: string): boolean {
  return ![MetricKey.releasability_rating, MetricKey.releasability_effort].includes(
    metric as MetricKey,
  );
}

export function hasTree(metric: string): boolean {
  return metric !== MetricKey.alert_status;
}

export function hasTreemap(metric: string, type: string): boolean {
  return (
    [MetricType.Percent, MetricType.Rating, MetricType.Level].includes(type as MetricType) &&
    hasTree(metric)
  );
}

export function hasBubbleChart(bubblesByDomain: BubblesByDomain, domainName: string): boolean {
  return bubblesByDomain[domainName] !== undefined;
}

export function hasFacetStat(metric: string): boolean {
  return metric !== MetricKey.alert_status;
}

export function hasFullMeasures(branch?: BranchLike) {
  return !branch || isBranch(branch);
}

export function getMeasuresPageMetricKeys(metrics: Dict<Metric>, branch?: BranchLike) {
  const metricKeys = getDisplayMetrics(Object.values(metrics)).map((metric) => metric.key);

  if (isPullRequest(branch)) {
    return metricKeys.filter((key) => isDiffMetric(key));
  }

  return metricKeys;
}

export function getBubbleMetrics(
  bubblesByDomain: BubblesByDomain,
  domain: string,
  metrics: Dict<Metric>,
) {
  const conf = bubblesByDomain[domain];
  return {
    x: metrics[conf.x],
    y: metrics[conf.y],
    size: metrics[conf.size],
    colors: conf.colors && conf.colors.map((color) => metrics[color]),
  };
}

export function getBubbleYDomain(bubblesByDomain: BubblesByDomain, domain: string) {
  return bubblesByDomain[domain].yDomain;
}

export function isProjectOverview(metric: string) {
  return metric === PROJECT_OVERVIEW;
}

function parseView(metric: MetricKey, rawView?: string): MeasurePageView {
  const view = (parseAsString(rawView) || DEFAULT_VIEW) as MeasurePageView;
  if (!hasTree(metric)) {
    return MeasurePageView.list;
  } else if (view === MeasurePageView.list && !hasList(metric)) {
    return MeasurePageView.tree;
  }
  return view;
}

export interface Query {
  asc?: boolean;
  metric: string;
  selected?: string;
  view: MeasurePageView;
}

export const parseQuery = memoize((urlQuery: RawQuery): Query => {
  const parsedMetric = parseAsString<MetricKey>(urlQuery['metric']) || DEFAULT_METRIC;
  const metric = MEASURES_REDIRECTION[parsedMetric] ?? parsedMetric;
  return {
    metric,
    selected: parseAsString(urlQuery['selected']),
    view: parseView(metric, urlQuery['view']),
    asc: parseAsOptionalBoolean(urlQuery['asc']),
  };
});

export const serializeQuery = memoize((query: Query) => {
  return cleanQuery({
    metric: query.metric === DEFAULT_METRIC ? undefined : serializeString(query.metric),
    selected: serializeString(query.selected),
    view: query.view === DEFAULT_VIEW ? undefined : serializeString(query.view),
  });
});
