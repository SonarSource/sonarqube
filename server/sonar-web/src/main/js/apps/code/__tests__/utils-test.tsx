/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { ComponentQualifier } from '~sonar-aligned/types/component';
import { mockMainBranch, mockPullRequest } from '../../../helpers/mocks/branch-like';
import { getCodeMetrics, mostCommonPrefix } from '../utils';

describe('getCodeMetrics', () => {
  it('should return the right metrics for portfolios', () => {
    expect(getCodeMetrics(ComponentQualifier.Portfolio)).toMatchSnapshot();
    expect(
      getCodeMetrics(ComponentQualifier.Portfolio, undefined, { includeQGStatus: true }),
    ).toMatchSnapshot();
    expect(
      getCodeMetrics(ComponentQualifier.Portfolio, undefined, {
        includeQGStatus: true,
        newCode: true,
      }),
    ).toMatchSnapshot();
    expect(
      getCodeMetrics(ComponentQualifier.Portfolio, undefined, {
        includeQGStatus: true,
        newCode: false,
      }),
    ).toMatchSnapshot();
  });

  it('should return the right metrics for apps', () => {
    expect(getCodeMetrics(ComponentQualifier.Application)).toMatchSnapshot();
  });

  it('should return the right metrics for projects', () => {
    expect(getCodeMetrics(ComponentQualifier.Project, mockMainBranch())).toMatchSnapshot();
    expect(getCodeMetrics(ComponentQualifier.Project, mockPullRequest())).toMatchSnapshot();
  });
});

describe('#mostCommonPrefix', () => {
  it('should correctly find the common path prefix', () => {
    expect(mostCommonPrefix(['src/main/ts/tests', 'src/main/java/tests'])).toEqual('src/main/');
    expect(mostCommonPrefix(['src/main/ts/app', 'src/main/ts/app'])).toEqual('src/main/ts/');
    expect(mostCommonPrefix(['src/main/ts', 'lib/main/ts'])).toEqual('');
  });
});
