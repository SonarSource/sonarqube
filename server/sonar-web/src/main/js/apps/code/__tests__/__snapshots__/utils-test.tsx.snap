// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCodeMetrics should return the right metrics for apps 1`] = `
[
  "alert_status",
  "ncloc",
  "software_quality_security_issues",
  "software_quality_reliability_issues",
  "software_quality_maintainability_issues",
  "vulnerabilities",
  "bugs",
  "code_smells",
  "security_hotspots",
  "coverage",
  "duplicated_lines_density",
]
`;

exports[`getCodeMetrics should return the right metrics for portfolios 1`] = `
[
  "releasability_rating",
  "new_security_rating",
  "new_software_quality_security_rating",
  "new_reliability_rating",
  "new_software_quality_reliability_rating",
  "new_maintainability_rating",
  "new_software_quality_maintainability_rating",
  "new_security_review_rating",
  "new_lines",
  "releasability_rating",
  "security_rating",
  "software_quality_security_rating",
  "reliability_rating",
  "software_quality_reliability_rating",
  "sqale_rating",
  "software_quality_maintainability_rating",
  "security_review_rating",
  "ncloc",
]
`;

exports[`getCodeMetrics should return the right metrics for portfolios 2`] = `
[
  "releasability_rating",
  "new_security_rating",
  "new_software_quality_security_rating",
  "new_reliability_rating",
  "new_software_quality_reliability_rating",
  "new_maintainability_rating",
  "new_software_quality_maintainability_rating",
  "new_security_review_rating",
  "new_lines",
  "releasability_rating",
  "security_rating",
  "software_quality_security_rating",
  "reliability_rating",
  "software_quality_reliability_rating",
  "sqale_rating",
  "software_quality_maintainability_rating",
  "security_review_rating",
  "ncloc",
  "alert_status",
]
`;

exports[`getCodeMetrics should return the right metrics for portfolios 3`] = `
[
  "releasability_rating",
  "new_security_rating",
  "new_software_quality_security_rating",
  "new_reliability_rating",
  "new_software_quality_reliability_rating",
  "new_maintainability_rating",
  "new_software_quality_maintainability_rating",
  "new_security_review_rating",
  "new_lines",
  "alert_status",
]
`;

exports[`getCodeMetrics should return the right metrics for portfolios 4`] = `
[
  "releasability_rating",
  "security_rating",
  "software_quality_security_rating",
  "reliability_rating",
  "software_quality_reliability_rating",
  "sqale_rating",
  "software_quality_maintainability_rating",
  "security_review_rating",
  "ncloc",
  "alert_status",
]
`;

exports[`getCodeMetrics should return the right metrics for projects 1`] = `
[
  "ncloc",
  "software_quality_security_issues",
  "software_quality_reliability_issues",
  "software_quality_maintainability_issues",
  "vulnerabilities",
  "bugs",
  "code_smells",
  "security_hotspots",
  "coverage",
  "duplicated_lines_density",
]
`;

exports[`getCodeMetrics should return the right metrics for projects 2`] = `
[
  "new_lines",
  "software_quality_security_issues",
  "software_quality_reliability_issues",
  "software_quality_maintainability_issues",
  "vulnerabilities",
  "bugs",
  "code_smells",
  "security_hotspots",
  "new_coverage",
  "new_duplicated_lines_density",
]
`;
